File: examples/checks/ev_roundtrip_planner.py
# Independent Python checks for the ev_roundtrip_planner example.
from .common import run_fragment_checks

CHECKS = [
    ('8 acceptable Brussels-to-Cologne plans were derived', ['acceptable plans : 8']),
    ('selected plan duration 210.0 is below 260.0', ['The selected plan is the fastest acceptable candidate under belief > 0.93, cost < 0.090, and duration < 260.0.']),
    ('selected plan cost 0.054 is below 0.090', ['cost : 0.054']),
    ('selected plan belief 0.974175 is above 0.93', ['It uses the shuttle from Aachen to Cologne, avoiding an extra charge stop while keeping belief at 0.974175.']),
    ('selected plan reaches Cologne', ['The planner starts with car1 at Brussels, battery=high, pass=none, then composes action descriptions until the goal city Cologne is reached.']),
    ('selected plan uses the high-belief Aachen-Cologne shuttle for the last mile', ['Duration and cost are summed across each candidate; belief and comfort are multiplied, matching the N3 planner pattern.']),
    ('bounded search consumed at most 8 of 8 fuel tokens', ['fuel remaining : 5 of 8']),
]

def run(ctx):
    return run_fragment_checks(ctx, CHECKS)

File: examples/checks/ev_roundtrip_planner.go
package main

import (
	"strings"
)

func checkEVRoundtrip(ctx *Context) []Check {
	d := ctx.M()
	plans, depth := evSearch(d)
	best := plans[0]
	ansPlan := []string{}
	if m := reFind(ctx.Answer, `Select plan : ([^.]+)\.`); m != nil {
		for _, p := range strings.Split(m[0], "->") {
			ansPlan = append(ansPlan, strings.TrimSpace(p))
		}
	}
	return []Check{{"bounded search finds eight acceptable Brussels-to-Cologne plans", len(plans) == 8 && grabInt(ctx.Answer, `acceptable plans : (\d+)`) == 8}, {"the selected plan is the fastest acceptable candidate", sliceEq(ansPlan, best.actions) && sliceEq(best.actions, []string{"drive_bru_liege", "drive_liege_aachen", "shuttle_aachen_cologne"})}, {"selected duration and fuel remaining are recomputed", close(fieldFloat(ctx.Answer, "duration"), best.duration, 5e-7) && grabInt(ctx.Answer, `fuel remaining : (\d+) of`) == best.fuel}, {"selected cost, belief, and comfort are recomputed by summing/multiplying actions", close(fieldFloat(ctx.Answer, "cost"), best.cost, 5e-4) && close(fieldFloat(ctx.Answer, "belief"), best.belief, 5e-7) && close(fieldFloat(ctx.Answer, "comfort"), best.comfort, 5e-7)}, {"the selected final state satisfies the wildcard goal", evMatches(asMap(d["Goal"]), best.state) && evKey(best.state) == "Cologne|low|none"}, {"the selected plan satisfies reliability, cost, and duration thresholds", best.belief > num(asMap(d["Thresholds"])["MinBelief"]) && best.cost < num(asMap(d["Thresholds"])["MaxCost"]) && best.duration < num(asMap(d["Thresholds"])["MaxDuration"])}, {"the last mile uses the high-belief Aachen-Cologne shuttle", best.actions[len(best.actions)-1] == "shuttle_aachen_cologne"}, {"search depth stays within the fuel-step bound", depth <= integer(d["FuelSteps"])}, {"the top two acceptable plans are ordered by duration then cost", plans[0].duration <= plans[1].duration && plans[0].cost < plans[1].cost}}
}
